src/test/java/io/reactivex/internal/operators/flowable/FlowableFromAsyncTest.java

```diff
@@ -0,0 +1,729 @@
+package io.reactivex.internal.operators.flowable;
+
+import org.junit.*;
+import org.reactivestreams.*;
+
+import io.reactivex.*;
+import io.reactivex.exceptions.*;
+import io.reactivex.functions.Consumer;
+import io.reactivex.processors.PublishProcessor;
+import io.reactivex.subscribers.*;
+
+public class FlowableFromAsyncTest {
+
+    PublishAsyncEmitter source;
+
+    PublishAsyncEmitterNoCancel sourceNoCancel;
+
+    TestSubscriber<Integer> ts;
+    
+    @Before
+    public void before() {
+        source = new PublishAsyncEmitter();
+        sourceNoCancel = new PublishAsyncEmitterNoCancel();
+        ts = new TestSubscriber<Integer>(0L);
+    }
+    
+    @Test
+    public void normalBuffered() {
+        Flowable.fromAsync(source, AsyncEmitter.BackpressureMode.BUFFER).subscribe(ts);
+        
+        source.onNext(1);
+        source.onNext(2);
+        source.onComplete();
+        
+        ts.request(1);
+        
+        ts.assertValue(1);
+        
+        Assert.assertEquals(0, source.requested());
+        
+        ts.request(1);
+        
+        ts.assertValues(1, 2);
+        ts.assertNoErrors();
+        ts.assertComplete();
+    }
+
+    @Test
+    public void normalDrop() {
+        Flowable.fromAsync(source, AsyncEmitter.BackpressureMode.DROP).subscribe(ts);
+        
+        source.onNext(1);
+
+        ts.request(1);
+
+        source.onNext(2);
+        source.onComplete();
+        
+        ts.assertValues(2);
```

Comment:
I mean, if possible better to check values than their count, because
operator may emit them in wrong order or one item twice, etc and your test
will pass.

On Tue, 12 Jul 2016, 19:26 David Karnok, <notifications@github.com> wrote:

> In
> src/test/java/io/reactivex/internal/operators/flowable/FlowableFromAsyncTest.java
> <https://github.com/ReactiveX/RxJava/pull/4192#discussion_r70472113>:
>
> > +        ts.assertNoErrors();
> > +        ts.assertComplete();
> > +    }
> > +
> > +    @Test
> > +    public void normalDrop() {
> > +        Flowable.fromAsync(source, AsyncEmitter.BackpressureMode.DROP).subscribe(ts);
> > +
> > +        source.onNext(1);
> > +
> > +        ts.request(1);
> > +
> > +        source.onNext(2);
> > +        source.onComplete();
> > +
> > +        ts.assertValues(2);
>
> What do you mean?
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ReactiveX/RxJava/pull/4192/files/7198f1afd376d6ef9375b72251aa50148d72d712#r70472113>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AA7B3LHYcCijs5dLw606DqUm8dn2TgiCks5qU8AjgaJpZM4JKKBF>
> .
>